rcl/src/rcl/introspection.c

```diff
@@ -0,0 +1,383 @@
+  strcpy(introspection_utils->service_type_name, service_type_name);
+
+  rcl_service_typesupport_to_message_typesupport(
+    service_type_support, &introspection_utils->request_type_support,
+    &introspection_utils->response_type_support, allocator);
+
+  // Make a publisher
+  char * service_event_topic_name = (char *)allocator->zero_allocate(
+    strlen(service_name) + strlen(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) + 1, sizeof(char),
+    allocator->state);
+  strcpy(service_event_topic_name, service_name);
+  strcat(service_event_topic_name, RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX);
+
+  introspection_utils->service_event_topic_name =
+    allocator->allocate(strlen(service_event_topic_name) + 1, allocator->state);
+
+  strcpy(introspection_utils->service_event_topic_name, service_event_topic_name);
+
+  introspection_utils->publisher = allocator->allocate(sizeof(rcl_publisher_t), allocator->state);
+  *introspection_utils->publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * service_event_typesupport =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(rcl_interfaces, msg, ServiceEvent);
+
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  ret = rcl_publisher_init(
+    introspection_utils->publisher, node, service_event_typesupport, service_event_topic_name,
+    &publisher_options);
+  if (RCL_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(rcl_get_error_string().str);
+    return RCL_RET_ERROR;
+  }
+
+  // make a clock
+  introspection_utils->clock = clock;
+  allocator->deallocate(service_event_topic_name, allocator->state);
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rcl_service_introspection_fini(
+  rcl_service_introspection_utils_t * introspection_utils, rcl_allocator_t * allocator,
+  rcl_node_t * node)
+{
+  rcl_ret_t ret;
+  ret = rcl_publisher_fini(introspection_utils->publisher, node);
+  if (RCL_RET_OK != ret) {
+    RCL_SET_ERROR_MSG(rcl_get_error_string().str);
+    return RCL_RET_ERROR;
+  }
+
+  allocator->deallocate(introspection_utils->publisher, allocator->state);
+  allocator->deallocate(introspection_utils->service_name, allocator->state);
+  allocator->deallocate(introspection_utils->service_event_topic_name, allocator->state);
+  allocator->deallocate(introspection_utils->service_type_name, allocator->state);
+
+  introspection_utils->publisher = NULL;
+  introspection_utils->service_name = NULL;
+  introspection_utils->service_event_topic_name = NULL;
+  introspection_utils->service_type_name = NULL;
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rcl_introspection_send_message(
+  const rcl_service_introspection_utils_t * introspection_utils, const uint8_t event_type,
+  const void * ros_response_request, const int64_t sequence_number, const uint8_t uuid[16],
+  const rcl_allocator_t * allocator)
+{
+  // Early exit of service introspection if it isn't enabled
+  // TODO(ihasdapie): Different return code?
+  if (!introspection_utils->_enabled) return RCL_RET_OK;
```

Comment:
I'm using this check to turn service introspection on/off during runtime, so it wouldn't make sense to have it be an error code. I think RCL_RET_OK is the right choice here because an early return without doing anything is expected.